cli/src/commands/run.rs

```diff
@@ -14,12 +14,262 @@
+    match id {
+        MergedTreeId::Legacy(tree_id) => tree_id.hex(),
+        MergedTreeId::Merge(tree_ids) => {
+            let ids = tree_ids
+                .map(|id| id.hex())
+                .iter_mut()
+                .enumerate()
+                .map(|(i, s)| {
+                    // Incredibly "smart" way to say, append "-" if the number is odd "+"
+                    // otherwise.
+                    if i & 1 != 0 {
+                        s.push('-');
+                    } else {
+                        s.push('+');
+                    }
+                    s.to_owned()
+                })
+                .collect_vec();
+            let mut obfuscated: String = ids.concat();
+            // `PATH_MAX` could be a problem for different operating systems, so truncate
+            // it.
+            if obfuscated.len() >= 255 {
+                obfuscated.truncate(200);
+            }
+            obfuscated
+        }
+    }
+}
+
+fn get_runtime(jobs: usize) -> tokio::runtime::Handle {
+    let mut builder = Builder::new_multi_thread();
+    if cfg!(watchman) {
+        // Watchman requires a multithreaded runtime, so just reuse it.
+        return runtime::Handle::current();
+    }
+    if jobs == 1 {
+        builder.max_blocking_threads(1);
+    } else {
+        builder.max_blocking_threads(jobs);
+    }
+    let rt = builder.build().unwrap();
+    rt.handle().clone()
+}
+
+/// A commit stored under `.jj/run/`
+// TODO: Create a caching backend, which creates these on a dedicated thread or
+// threadpool.
+struct StoredCommit {
+    /// Obfuscated name for an easier lookup. If a tree/directory its not set
+    name: Option<String>,
+    /// The respective commit unmodified.
+    commit: Commit,
+    /// The output directory of the commit, contains stdout and stderr for it
+    output_dir: PathBuf,
+    /// Self-explanatory
+    working_copy_dir: PathBuf,
+    state_dir: PathBuf,
+    /// The `stdout` of the commit
+    stdout: File,
+    /// The `stderr` of the commit
+    stderr: File,
+}
+
+impl StoredCommit {
+    fn new(
+        name: Option<String>,
+        commit: &Commit,
+        output_dir: PathBuf,
+        working_copy_dir: PathBuf,
+        state_dir: PathBuf,
+        stdout: File,
+        stderr: File,
+    ) -> Self {
+        Self {
+            name,
+            commit: commit.clone(),
+            output_dir,
+            working_copy_dir,
+            state_dir,
+            stdout,
+            stderr,
+        }
+    }
+}
+
+fn create_output_files(id: CommitId, path: &PathBuf) -> Result<(File, File), io::Error> {
+    let stdout_path = path.join("output").join(format!("stdout.{id}"));
+    let stderr_path = path.join("output").join(format!("stderr.{id}"));
+    let mut file_options = OpenOptions::new();
+    let stdout = file_options.write(true).open(stdout_path)?;
+    let stderr = file_options.write(true).open(stderr_path)?;
+    Ok((stdout, stderr))
+}
+
+fn create_working_copies(
```

Comment:
suggestion: I would not preemptively create the working-copies; instead, I would invoke a function to create a working-copy on demand as part of running the job itself.

- We may as well benefit from the parallelization that `run` itself would use, rather than having to write it in two places.
- Extensions to `run` might make it so that we don't actually use all the working copies we allocated up-front.

Reply:
Ack, but let's first get it running somewhat.

Reply:
I'd still defer this unitl some work from @hooper has happened.

Reply:
A lot of this will shake out when we devise the abstractions for varying parallelism, execution order, caching, etc. This PR is probably too early to put much effort into that.

I think working copy creation will ultimately be triggered by the part that determines parallelism and execution order, so that we can hide the latency sometimes while also deferring the work until it is almost needed. Having it in the inner loop may make sense until then, although I wonder if the parallel access pattern would make it slower.

Reply:
> I think working copy creation will ultimately be triggered by the part that determines parallelism and execution order, so that we can hide the latency sometimes while also deferring the work until it is almost needed.

+1 from me for the general plan. 

> Having it in the inner loop may make sense until then, although I wonder if the parallel access pattern would make it slower.

I think a fan-out option makes sense if everything is accessed through the network, this also wouldn't hurt the planned on-disk implementation as much.